src/metabase/models/params/card_values.clj

```diff
@@ -14,44 +15,82 @@
+                            (and (= (:name field-info)
+                                    id-or-name)
+                                 (= (:base-type options)
+                                    (:base-type field-info))))
+                        result-metadata)))))
+
 (defn- values-from-card-query
-  [card-id value-field query]
-  (let [card        (db/select-one Card :id card-id)]
-    {:database   (:database_id card)
-     :type       :query
-     :query      (merge
-                  {:source-table (format "card__%d" card-id)
-                   :fields       [value-field]
-                   :limit        *max-rows*}
-                  (when query
-                    {:filter [:contains [:lower value-field] (u/lower-case-en query)]}))
-     :middleware {:disable-remaps? true}}))
-
-(s/defn values-from-card
-  "Get values a field from a card.
+  [card value-field query]
+  (let [value-base-type (:base_type (field->field-info value-field (:result_metadata card)))]
+     {:database (:database_id card)
+      :type     :query
+      :query    (merge
+                   {:source-table (format "card__%d" (:id card))
+                    :breakout     [value-field]
+                    :limit        *max-rows*}
+                   {:filter [:and
+                             [(if-not (isa? value-base-type :type/Text)
+                                 :not-null
+                                 :not-empty)
+                              value-field]
+                             (when query
+                                (if-not (isa? value-base-type :type/Text)
+                                   [:= value-field query]
+                                   [:contains [:lower value-field] (u/lower-case-en query)]))]})
+      :middleware {:disable-remaps? true}}))
+
+(mu/defn values-from-card
+  "Get distinct values of a field from a card.
 
   (values-from-card 1 [:field \"name\" nil] \"red\")
   ;; will execute a mbql that looks like
   ;; {:source-table (format \"card__%d\" card-id)
   ;;  :fields       [value-field]
+  ;;  :breakout     [value-field]
   ;;  :filter       [:contains [:lower value-field] \"red\"]
   ;;  :limit        *max-rows*}
   =>
   {:values          [\"Red Medicine\"]
    :has_more_values false}
   "
-  ([card-id value-field-ref]
-   (values-from-card card-id value-field-ref nil))
-
-  ([card-id     :- su/IntGreaterThanZero
-    value-field :- su/Field
-    query       :- (s/maybe su/NonBlankString)]
-   (let [mbql-query   (values-from-card-query card-id value-field query)
-         query-limit  (get-in mbql-query [:query :limit])
-         result       (qp/process-query mbql-query)]
-     {:values          (map first (get-in result [:data :rows]))
+  ([card-id value-field]
+   (values-from-card card-id value-field nil))
+
+  ([card-id         :- ms/IntGreaterThanZero
+    value-field     :- ms/Field
+    query           :- [:any]]
+   (let [card         (db/select-one Card :id card-id)
```

Comment:
I think we should do an existence check on the `card` here. In case a card is deleted the error message from a parameter would be strange. Should we also check on archive status?

Reply:
good point, I added checks for card existence and archive [here](https://github.com/metabase/metabase/pull/27697/commits/b84d64f005a8e00702dd28f9eaf41ebdf224cf66).

We non-string filter, we filter using exact match (`=` instead of `contains`), so I don't think we need to catch that.